### PR TITLE
Integration of modes

### PR DESCRIPTION
**Pull request by glennclarence**

### CODE REVIEW

File: .settings/language.settings.xml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (UmbrellaSampler) [inline .settings/language.settings.xml:1]: I think you added the whole .settings folder. This must be removed from the git index. You can google how to remove a committed folder. Then add the .settings folder to the .gitignore file.

File: src/apps/sc/scATTRACT.tpp [hunk 2]
```diff
@@ -43,7 +44,7 @@ void scATTRACT<SERVICE>::run() {
 	server->wait(request, results.data());
 
 	for (result_t const res : results) {
-		std::cout << res << std::endl;
+	//	std::cout << res << std::endl;
```
Comment (UmbrellaSampler) [inline src/apps/sc/scATTRACT.tpp]: Why is this line commented? Shouldn't we print the results?

File: src/cli/TwoBodyParser.cpp [hunk 3]
```diff
@@ -47,6 +49,7 @@ void TwoBodyParser::addOptions() noexcept {
 	_optsDesc.add(concurrency);
 	po::options_description sim("simulation");
 	sim.add_options()
+			("numModes", po::value<int>()->default_value(DEFAULT_NUM_MODES), "number of modes")
```
Comment (UmbrellaSampler) [inline src/cli/TwoBodyParser.cpp]: What is "numModes"? is it the number of modes we want to use even if more modes are available? What happens if someone sets is to --numModes 100? What happens if someone provides a mode file and forgets to specify --numModes? I suggest to use all that are given in the files

File: src/cli/TwoBodyParser.cpp [hunk 4]
```diff
@@ -84,6 +87,10 @@ void TwoBodyParser::assigneArgs(po::variables_map const& vm) noexcept {
 		_args->paramsName = vm["par"].as<string>();
 	if(vm.count("alphabet"))
 		_args->alphabetName = vm["alphabet"].as<string>();
+	if(vm.count("modr"))
+				_args->recModesName = vm["modr"].as<string>();
```
Comment (UmbrellaSampler) [inline src/cli/TwoBodyParser.cpp]: too much tabs?

File: src/cli/parser_constants.h [hunk 5]
```diff
@@ -24,12 +24,14 @@ constexpr auto DEFAULT_LIGANG_PDB_FILE = "ligandr.pdb";
...
 constexpr auto DEFAULT_GRID_ALPAHBET_FILE = "receptorgrid.alphabet";
+constexpr auto DEFAULT_MODE_RECEPTOR_FILE = "modesReceptor.dat";
+constexpr auto DEFAULT_MODE_LIGAND_FILE = "modesLigand.dat";
 
 /**
  * Server context
  */
 constexpr int DEFAULT_NUM_CPUS = 1;
-
+constexpr int DEFAULT_NUM_MODES = 0;
```
Comment (UmbrellaSampler) [inline src/cli/parser_constants.h]: is that really a good default value if someone wants to use modes?

File: src/apps/AppFactory.cpp [hunk 6]
```diff
@@ -10,6 +10,7 @@
 #include "scATTRACT.h"
```
Comment (UmbrellaSampler) [inline src/apps/AppFactory.cpp]: I think I already refactored this AppFactory. resolve your conflicts first and then make a pull request.

File: src/apps/AppFactory.h [hunk 7]
```diff
@@ -26,6 +26,7 @@ class AppFactory {
 	enum class Platform {
 		CPU,
 		GPU,
+		CPUModes,
```
Comment (UmbrellaSampler) [inline src/apps/AppFactory.h]: This Platform enum does not exist anymore. Again, first resolve your confilcts with the master

File: src/apps/sc/scATTRACT.cpp [hunk 8]
```diff
@@ -20,6 +21,12 @@ class scATTRACT<CPUEnergyService6D<float>>;
 template
 class scATTRACT<CPUEnergyService6D<double>>;
```
Comment (UmbrellaSampler) [inline src/apps/sc/scATTRACT.cpp]: There is no CUDA awareness in this file anymore in the current version of the master. Conflicts!!

File: src/model/Types_6D_Modes.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+#endif
+
+template<typename REAL>
+struct DOF_6D_Modes {
+	using real_t = typename TypeWrapper<REAL>::real_t;
+	using vec3_t = Vec3<real_t>;
+	vec3_t pos;
+	vec3_t ang;
+	real_t modes[10];
```
Comment (UmbrellaSampler) [inline src/model/Types_6D_Modes.h]: better don't hard-code the "10" for the number of modes. For this kind of things I prefer to the create a configuration file (config.h ?!) where you define the maximum allowed number of modes. This file can be included in any modules that need that information. Thus we are more flexible, if we want to change the number later...

File: src/model/reduction.h [hunk 10]
```diff
@@ -27,6 +28,7 @@ class PotForce {
 public:
 	real_t E;
 	vec3_t pos;
+	real_t modes[10];
```
Comment (UmbrellaSampler) [inline src/model/reduction.h]: see comment in ""Types_6d_Modes.h": do not hard-code the "10"
I think we need to distinguish between a PotForce class as it was, and a new "PotForceModes". Maybe a new file for those?

File: src/service/CPUEnergyService6D.tpp [hunk 11]
```diff
@@ -170,6 +170,18 @@ auto CPUEnergyService6D<REAL>::createItemProcessor() -> itemProcessor_t {
 					lig->numAtoms()
 			); // OK
 
+			reduceModeForce(
```
Comment (UmbrellaSampler) [inline src/service/CPUEnergyService6D.tpp]: this function shouldn't be in this file (but in CPUEnergyService6DModes.tpp)
)

File: src/service/Configurator_6D_Modes.tpp [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+#include "nativeTypesMath.h"
+#include "ServiceFactory.h"
+
+
+namespace as {
+
+template<typename SERVICE>
+void Configurator_6D_Modes<SERVICE>::init(CmdArgs const& args) noexcept {
+
+	/* load dataItems */
```
Comment (UmbrellaSampler) [inline src/service/Configurator_6D_Modes.tpp:29]: Ok for now. But this seems to be a lot of code duplication (compare Configurator_6D.tpp). We need to overcome this soon. Can you create a ticket for that?!

File: test/test_HMMread.h [hunk 13]
```diff
@@ -0,0 +1,156 @@
+/*
```
Comment (UmbrellaSampler) [inline test/test_HMMread.h:1]: What is this file for? Didn't you implement that already in readFile.cpp?

**Review by UmbrellaSampler — CHANGES_REQUESTED** [review]
First of all: before you do a pull request, please merge the changes that already affected the master (I did changes in the meanwhile) into your feature branch (locally). Then push again. I will not comment again on a pull request that have merge conflicts. Do you know what I mean?
   
Maybe try to do less changes within one pull request. By doing so, you can avoid lots of merge conflicts.

The tests you wrote are a bit lengthy. It seems that you reimplemented the read functionality, didn't you? Furthermore, you don't make use of the Testing Framework to check for equality! It comes with lots of stuff for checking equality and handling failed assertions. I think I need to introduce you a little bit more into how unit tests should look like. I simple example is Server_Test.cpp (does it still compile?). 
However, I take the test sources as they are...

File: test/main.cpp [hunk 14]
```diff
@@ -6,10 +6,10 @@
  */
 
 #include "gtest/gtest.h"
-
-//int main(int argc, char **argv)
-//{
-//    testing::InitGoogleTest(&argc, argv);
-//    return RUN_ALL_TESTS();
-//}
+#include "test_HMMread.h"
```
Comment (UmbrellaSampler) [inline test/main.cpp]: this include statement should not be here

File: .settings/language.settings.xml (hunk 1 above)
Comment (glennclarence) [reply .settings/language.settings.xml:1]: danke werde ich mir anschauen! Habe die ganze files irgendwann immer manually vom commit ausgeschlossen.

File: src/apps/sc/scATTRACT.tpp (hunk 2 above)
Comment (glennclarence) [reply src/apps/sc/scATTRACT.tpp]: totally right. komisches artefakt

File: src/model/Types_6D_Modes.h (hunk 9 above)
Comment (glennclarence) [reply src/model/Types_6D_Modes.h]: war mir auch ein bischen unangenehm das so zu machen. Die Lösung kling gut, ich werde es einbauen. 
Eine alternative die mir eingefallen ist, war die Anzahl der Moden beim erstellen einer DOF struct zu initialisieren. Was hälst du davon?

File: src/service/CPUEnergyService6D.tpp (hunk 11 above)
Comment (glennclarence) [reply src/service/CPUEnergyService6D.tpp]: true

File: src/model/reduction.h (hunk 10 above)
Comment (glennclarence) [reply src/model/reduction.h]: was hälst du davon datei potforcetypes.h zu erstellen und eine reduction_modes.h. 

die potforcetypes.h enthält die types für und ohne moden und die reduction_modes.h hat die included reduction.h und enthält noch alle funktionen die unique für die mode reduction sind. 

So hätte man die moden und nicht moden klar getrennt und doppelten code vermieden.

File: src/service/Configurator_6D_Modes.tpp (hunk 12 above)
Comment (glennclarence) [reply src/service/Configurator_6D_Modes.tpp:29]: done

File: test/test_HMMread.h (hunk 13 above)
Comment (glennclarence) [reply test/test_HMMread.h:1]: ja merkwürdig eigentlich war da was drin..

Comment (glennclarence) [reply test/test_HMMread.h:1]: habe das file verwechselt. Die Moden werden in Attract noch einmal verändert (auf den gesamtbetrag der Moden normiert) dieser Test ist dafür gedacht diesen Teil der eigentlichen readHMM funktion zu überprüfen.

**Review by UmbrellaSampler — APPROVED** [review]

**Comment by UmbrellaSampler** [comment]
gut gemacht!

**Pull request merged**